gs_chunked_io/reader.py

```diff
@@ -97,45 +95,37 @@ def write(self, *args, **kwargs):
         raise OSError()
 
     def close(self):
-        if self.executor:
-            self.executor.shutdown()
         super().close()
 
-def for_each_chunk(blob: Blob, chunk_size: int=default_chunk_size, threads: Optional[int]=2):
+def for_each_chunk(blob: Blob, chunk_size: int=default_chunk_size, async_queue: Optional[AsyncQueue]=None):
     """
-    Fetch chunks and yield in order. Chunks are downloaded with concurrency equal to `threads`
+    Fetch chunks and yield in order. Chunks are downloaded with concurrency as configured in `async_queue`
     """
     reader = Reader(blob, chunk_size=chunk_size)
-    if threads is not None:
-        assert 1 <= threads
-        with ThreadPoolExecutor(max_workers=threads) as e:
-            future_chunk_downloads = AsyncQueue(e, concurrency=threads)
-            for chunk_number in reader._unfetched_chunks:
-                future_chunk_downloads.put(reader._fetch_chunk, chunk_number)
-            for chunk in future_chunk_downloads.consume():
-                yield chunk
+    if async_queue is not None:
+        for chunk_number in reader._unfetched_chunks:
+            async_queue.put(reader._fetch_chunk, chunk_number)
+        for chunk in async_queue.consume():
+            yield chunk
     else:
         for chunk_number in reader._unfetched_chunks:
             yield reader._fetch_chunk(chunk_number)
 
-def for_each_chunk_async(blob: Blob, chunk_size: int=default_chunk_size, threads: int=4):
+def for_each_chunk_async(blob: Blob, async_set: AsyncSet, chunk_size: int=default_chunk_size):
     """
-    Fetch chunks with concurrency equal to `threads`, yielding results as soon as available.
+    Fetch chunks with concurrency as configured in `async_set`, yielding results as soon as available.
     Results may be returned in any order.
     """
-    assert 1 <= threads
     reader = Reader(blob, chunk_size)
 
     def fetch_chunk(chunk_number):
         data = reader._fetch_chunk(chunk_number)
         return chunk_number, data
 
-    with ThreadPoolExecutor(max_workers=threads) as e:
-        chunks = AsyncSet(e, threads)
-        for chunk_number in range(reader.number_of_chunks):
-            for chunk in chunks.consume_finished():
-                yield chunk
-                break
-            chunks.put(fetch_chunk, chunk_number)
-        for chunk in chunks.consume():
+    for chunk_number in range(reader.number_of_chunks):
+        for chunk in async_set.consume_finished():
             yield chunk
+            break
```

Comment:
It's not immediately obvious to me why there's a break here.  Could a comment be added?

Reply:
will fix

Reply:
Added comment:
># Breaking after the first yield allows us to add more downloads to the pot without           
># waiting for the client to complete potentially time-consuming operations.